src/dtx/dtx_resync.c

```diff
@@ -0,0 +1,387 @@
+		  struct dtx_resync_head *drh, int count, uint32_t version,
+		  bool block)
+{
+	struct dtx_resync_entry		*dre;
+	int				 rc = 0;
+	int				 i = 0;
+
+	D_ASSERT(drh->drh_count >= count);
+
+	/* If we are in block mode, then we need to commit the DTXs as fast as
+	 * possible, so just add them into the CoS cache, then the general DTX
+	 * commit mechanism will commit them some time later.
+	 */
+	if (block) {
+		int	err;
+
+		do {
+			dre = d_list_entry(drh->drh_list.next,
+					   struct dtx_resync_entry, dre_link);
+			if (dre->dre_hash == 0)
+				/* For punch object or the case that the dkey
+				 * hash is just zero, commit them immediately
+				 * without caching.
+				 */
+				err = dtx_commit(po_uuid, cont->sc_uuid,
+						 &dre->dre_dte, 1, version);
+			else
+				err = vos_dtx_add_cos(cont->sc_hdl,
+					&dre->dre_oid, &dre->dre_xid,
+					dre->dre_hash, dre->dre_intent ==
+					DAOS_INTENT_PUNCH ? true : false);
+			if (err != 0) {
+				D_ERROR("Failed to %s the DTX " DF_UOID"/"DF_DTI
+					": rc = %d\n",
+					dre->dre_hash == 0 ? "commit" : "cache",
+					DP_UOID(dre->dre_oid),
+					DP_DTI(&dre->dre_xid), err);
+				rc = err;
+			}
+			dtx_dre_release(drh, dre);
+		} while (++i < count);
+	} else {
+		struct daos_tx_entry	*dte = NULL;
+
+		D_ALLOC_ARRAY(dte, count);
+		if (dte == NULL)
+			return -DER_NOMEM;
+
+		do {
+			dre = d_list_entry(drh->drh_list.next,
+					   struct dtx_resync_entry, dre_link);
+			dte[i].dte_xid = dre->dre_xid;
+			dte[i].dte_oid = dre->dre_oid;
+			dtx_dre_release(drh, dre);
+		} while (++i < count);
+
+		rc = dtx_commit(po_uuid, cont->sc_uuid, dte, count, version);
+		if (rc < 0)
+			D_ERROR("Failed to commit the DTX: rc = %d\n", rc);
+		D_FREE(dte);
+	}
+
+	return rc > 0 ? 0 : rc;
+}
+
+static int
+dtx_status_handle(struct dtx_resync_args *dra, bool block)
+{
+	struct ds_cont			*cont = dra->cont;
+	struct pl_obj_layout		*layout = NULL;
+	struct dtx_resync_head		*drh = &dra->tables;
+	struct dtx_resync_entry		*dre;
+	struct dtx_resync_entry		*next;
+	int				 count = 0;
+	int				 err = 0;
+	int				 rc;
+
+	if (drh->drh_count == 0)
+		return 0;
+
+	d_list_for_each_entry_safe(dre, next, &drh->drh_list, dre_link) {
+		int	local_rc;
+
+		if (layout != NULL) {
+			pl_obj_layout_free(layout);
+			layout = NULL;
+		}
+
+		rc = ds_pool_check_leader(dra->po_uuid, &dre->dre_oid,
+					  dra->version, &layout);
+		if (rc <= 0) {
+			if (rc < 0)
+				D_WARN("Not sure about the leader for the DTX "
+				       DF_UOID"/"DF_DTI" (ver = %u): rc = %d, "
+				       "skip it.\n",
+				       DP_UOID(dre->dre_oid),
+				       DP_DTI(&dre->dre_xid), dra->version, rc);
+			else
+				D_DEBUG(DB_TRACE, "Not the leader for the DTX "
+					DF_UOID"/"DF_DTI" (ver = %u) skip it\n",
+					DP_UOID(dre->dre_oid),
+					DP_DTI(&dre->dre_xid), dra->version);
+			dtx_dre_release(drh, dre);
+			continue;
+		}
+
+		rc = dtx_check(dra->po_uuid, cont->sc_uuid,
+			       &dre->dre_dte, layout);
```

Comment:
We have ever considered to make batched DTX check. But that will make the logic some complex. Because different DTXs may belong to different redundancy group, we need to parse the replied status from multiple servers and recognize the replied status belong to which DTX, then combine them together carefully. That is not impossible, but I prefer to optimize them in next step.